paasta_tools/cli/cmds/mark_for_deployment.py

```diff
@@ -715,68 +748,106 @@ def status_code_by_state(self) -> Mapping[str, int]:
+
+        if self.old_git_sha != self.commit:
+            if self.state in self.rollback_states:
+                buttons.append("forward")
+            if self.state in self.rollforward_states:
+                buttons.append("rollback")
+
+        return buttons
+
+    def get_active_button(self) -> Optional[str]:
+        return {
+            'start_deploy': 'forward',
+            'deploying': 'forward',
+            'deployed': None,
+            'start_rollback': 'rollback',
+            'rolling_back': 'rollback',
+            'rolled_back': None,
+        }.get(self.state)
+
     def on_enter_mfd_failed(self):
-        self.available_buttons = []
         self.update_slack_status(f"Marking `{self.commit[:8]}` for deployment for {self.deploy_group} failed. Please see Jenkins for more output.")  # noqa E501
 
     def on_enter_deploying(self):
-        # if self.block is true, then deploying is a terminal state so we will promptly exit.
-        # Don't bother starting the background thread in this case.
-        self.active_button = "forward"
         self.update_slack()
+        # if self.block is False, then deploying is a terminal state so we will promptly exit.
+        # Don't bother starting the background thread in this case.
         if self.block:
-            thread = Thread(target=self.do_wait_for_deployment, args=(), daemon=True)
+            thread = Thread(target=self.do_wait_for_deployment, args=(self.commit,), daemon=True)
             thread.start()
 
     def on_exit_deploying(self):
         self.wait_for_deployment_green_light.clear()
-        self.update_slack_thread("Finished waiting for the deployment")
 
     def on_enter_start_rollback(self):
-        self.active_button = "rollback"
         self.update_slack_status(f"Rolling back ({self.deploy_group}) to {self.old_git_sha}")
-        mark_for_deployment(
+        self.mark_for_deployment_return_code = mark_for_deployment(
             git_url=self.git_url,
             deploy_group=self.deploy_group,
             service=self.service,
             commit=self.old_git_sha,
         )
-        self.available_buttons = []
-        self.update_slack_thread(f"Finished rollback mark-for-deployment to {self.old_git_sha}")
-        self.update_slack_status(f"Not waiting for rollback of {self.deploy_group} to {self.old_git_sha} to finish.")
-        self.retval = 1
 
-    def on_enter_deploy_aborted(self):
+        if self.mark_for_deployment_return_code != 0:
+            self.trigger('mfd_failed')
+        else:
+            self.update_slack_status(f"Marked `{self.old_git_sha[:8]}` for {self.deploy_group}.")
+            self.trigger('mfd_succeeded')
+
+    def on_enter_rolling_back(self):
+        self.update_slack()
+        if self.block:
+            thread = Thread(target=self.do_wait_for_deployment, args=(self.old_git_sha,), daemon=True)
+            thread.start()
+
+    def on_exit_rolling_back(self):
+        self.wait_for_deployment_green_light.clear()
```

Comment:
Is there a race condition between having the light cleared, switching from back or forward, and then having the light uncleared?

In other words, can we switch states fast enough for either of the wait_for_deployment threads from noticing? And does it matter?

Reply:
Yeah I was wondering about this. I couldn't get it to happen in practice, but that probably doesn't mean it's impossible. I guess there are two clean ways to deal with this:

1. use separate "green lights" for deployment and rollback
2. use a lock to make sure only one wait_for_deployment is happening at a time

I think I'm leaning towards option 1

Reply:
Hehe, if only we could kill threads!

Reply:
Can you take another look? I went with option 1.